### PR TITLE
feat(payment-processor): Batch p processor

### PR DESCRIPTION
**Pull request by olivier7delf**
## Create payment-processor Batch function
Using BatchProxy, from request's batch smart contract, it allows to do a batch transaction of ETH or ERC20 tokens.

It has to accept requests using Erc20FeeProxy, EthFeeProxy, and EthProxy.
(!) request denominated with EthProxy will use in batch payment: EthFeeProxy.

### CODE REVIEW

File: packages/payment-processor/test/payment/eth-batch-proxy.test.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    // check every balance: payer, feeAddress, payee1
+    expect(balanceEthBefore.toString()).toBe(
+      balanceEthAfter
+        .add(BigNumber.from(validRequest.expectedAmount).mul(2).toString()) // 2 is Nb_txs
+        .add('6') // = (2 + 1) * Nb_txs, and Nb_txs = 2 -> (fee + batchFee) * Nb_txs
+        .add(confirmedTx.gasUsed?.mul(tx?.gasPrice ?? 1))
+        .toString(),
+    );
+    expect(balanceFeeEthAfter.toString()).toBe(balanceFeeEthBefore.add('6').toString()); // = 2 * Nb_txs, and Nb_txs = 2
```
Comment (bertux) [inline packages/payment-processor/test/payment/eth-batch-proxy.test.ts:176]: 2 * 2 = 4 and not 6

Comment (olivier7delf) [reply packages/payment-processor/test/payment/eth-batch-proxy.test.ts]: done thanks!

**Review by alexandre-abrioux — APPROVED** [review]
LGTM, some minor suggestions but solid work!

File: packages/payment-processor/src/payment/utils.ts [hunk 2]
```diff
@@ -305,3 +305,23 @@ export function getAmountToPay(
...
+  request: ClientTypes.IRequestData,
+): void {
+  if (pn?.type !== getPaymentNetworkExtension(request)?.type) {
+    throw new Error(`Every payment network must be identical`);
+  }
+
+  if (pn?.version !== getPaymentNetworkExtension(request)?.version) {
+    throw new Error(`Every payment version must be identical`);
+  }
+}
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/utils.ts]: Thanks for the review! I updated it

File: packages/payment-processor/src/payment/batch-proxy.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+  }
+
+  return [
+    tokenAddresses,
+    paymentAddresses,
+    amountsToPay,
+    paymentReferences,
+    feesToPay,
+    feeAddressUsed,
+  ];
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-proxy.ts]: Thanks for the review! It s done

File: packages/payment-processor/test/payment/erc20-batch-proxy.test.ts [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+      const request = Utils.deepCopy(validRequest);
+      request.currencyInfo.value = '';
+      await expect(
+        payBatchProxyRequest([validRequest, request], batchVersion, wallet, batchFee),
+      ).rejects.toThrowError(
+        'request cannot be processed, or is not an pn-erc20-fee-proxy-contract request',
+      );
+    });
+
+    it('should throw an error if currencyInfo has no network', async () => {
```
Comment (alexandre-abrioux) [inline packages/payment-processor/test/payment/erc20-batch-proxy.test.ts:297]: You already have this same test in the suite above (I think), in `encodePayErc20BatchRequest`. Same comment for:
- 'should throw an error if the request is not erc20'
- 'should throw an error if the currencyInfo has no value'
- 'should throw an error if request has no extension'
- 'should throw an error if there is a wrong version mapping'

**Review by alexandre-abrioux — APPROVED** [review]
Nice! 😎 Some questions / suggestions but LGTM

File: packages/payment-processor/test/payment/erc20-batch-proxy.test.ts [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        expect(
+          BigNumber.from(feeBalanceErc20After).eq(
+            BigNumber.from(feeBalanceErc20Before).add((2 + 10) * 2),
+          ),
+        ).toBeTruthy();
+      });
+    });
+
+    describe('prepareBatchPaymentTransaction', () => {
+      it('should consider the version mapping', () => {
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/test/payment/erc20-batch-proxy.test.ts:420]: updated: I changed the second request to map

File: packages/payment-processor/test/payment/erc20-batch-proxy.test.ts (hunk 4 above)
Comment (olivier7delf) [reply packages/payment-processor/test/payment/erc20-batch-proxy.test.ts:297]: updated, now 2 different ERC20

File: packages/payment-processor/src/payment/utils.ts [hunk 6]
```diff
@@ -186,14 +186,21 @@ const currenciesMap: any = {
...
+  isEthBatch = false,
 ): void {
   const { feeAmount, feeAddress, expectedFlowRate, expectedStartDate } =
     getRequestPaymentValues(request);
-  const extension = request.extensions[paymentNetworkId];
+  let extension = request.extensions[paymentNetworkId];
+
+  if (isEthBatch && !extension) {
+    extension = request.extensions[PaymentTypes.PAYMENT_NETWORK_ID.ETH_INPUT_DATA];
+  }
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/src/payment/utils.ts:203]: It is done, it looks goods and all tests works, nice trick.
(I wait for the second reviewer before to resolve this one)

File: packages/payment-processor/test/payment/eth-batch-proxy.test.ts [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        .add(BigNumber.from(validRequest.expectedAmount).mul(2).toString()) // 2 is Nb_txs
+        .add('24') // = (2 + 10) * 2, and Nb_txs = 2 -> (fee + batchFee) * Nb_txs
+        .add(confirmedTx.gasUsed?.mul(tx?.gasPrice ?? 1))
+        .toString(),
+    );
+    expect(balanceFeeEthAfter.toString()).toBe(balanceFeeEthBefore.add('24').toString()); // = (fee+ batchFee) * Nb_txs <=> (2 + 10) * 2
+    expect(balancePayeeEthAfter1.toString()).toBe(balancePayeeEthBefore1.add('2000').toString()); // = 1000 * Nb_txs, and Nb_txs = 2
+  });
+
+  it('should pay an ETH batch of 2 requests with different receiverAddress and pay fee & batch fee and payment network id', async () => {
```
[the remark this answers is not in the document]
Comment (olivier7delf) [reply packages/payment-processor/test/payment/eth-batch-proxy.test.ts:188]: It s already the case in the case (or I misunderstood your comment)

File: packages/payment-processor/src/payment/utils.ts [hunk 8]
```diff
@@ -186,14 +186,19 @@ const currenciesMap: any = {
  * Utility to validate a request currency and payment details against a paymentNetwork.
  * @param request
  * @param paymentNetworkId
+ * FIXME: updating the extension: not needed anymore when "invoicing" will use only ethFeeProxy
```
Comment (alexandre-abrioux) [inline packages/payment-processor/src/payment/utils.ts:189]: Maybe bring this FIXME above the `if()`

File: packages/payment-processor/test/payment/erc20-batch-proxy.test.ts [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+            ],
+            batchVersion,
+            batchFee,
+          ).to,
+        ).toBe(batchPaymentsArtifact.getAddress('private', '0.1.0'));
+      });
+    });
+  });
+
+  describe('encodePayBatchRequest: multi tokens -> using two ERC20', () => {
```
Comment (alexandre-abrioux) [inline packages/payment-processor/test/payment/erc20-batch-proxy.test.ts:282]:
```suggestion
  describe('encodePayBatchRequest: multi tokens -> using two different ERC20', () => {
```

File: packages/payment-processor/test/payment/erc20-batch-proxy.test.ts [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+    it('should throw an error if the request is not erc20, for two ERC20', async () => {
```
Comment (alexandre-abrioux) [inline packages/payment-processor/test/payment/erc20-batch-proxy.test.ts:283]: There is a lot of repetition in your tests. Suggestion: you could make it shorter by doing something like bellow.
This is just a quick example but you could adapt this as you wish if you want to factorize your test suite.
```ts
const testSuite = (
  suiteName: string,
  requestTemplate1: ClientTypes.IRequestData,
  requestTemplate2: ClientTypes.IRequestData
) => {
  describe(suiteName, () => {
    let request1: ClientTypes.IRequestData;
    let request2: ClientTypes.IRequestData;
   
    beforeEach(() => {
      request1 = Utils.deepCopy(requestTemplate1) as ClientTypes.IRequestData;
      request2 = Utils.deepCopy(requestTemplate2) as ClientTypes.IRequestData;
    });

    it('test1', ...) // write your tests here directly using request1 and request2
    it('test2', ...)
  })
}

describe('erc20-batch-proxy', () => {
  testSuite('encodePayErc20BatchRequest for one type of ERC20', validRequest, validRequest);
  testSuite('encodePayErc20BatchRequest using two different ERC20', validRequest, fauValidRequest);
});
```

**Review by alexandre-abrioux — APPROVED** [review]

File: packages/payment-processor/test/payment/eth-batch-proxy.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexandre-abrioux) [reply packages/payment-processor/test/payment/eth-batch-proxy.test.ts]: Yes! Didn't see it 👍

**Review by leoslr — APPROVED** [review]

**Pull request merged**